### PR TITLE
Fix eyebot spawning

### PR DESCRIPTION
**Pull request by drbig**
Simplest fix for the eyebot spawning mechanic:
- For smashing-triggered alarms first check if the player was nearby
- Spawn at most one eyebot per trigger (but now the spawning chance is 2x higher)

The eyebots still spawn using the current logic (i.e. near the player, wherever she/he is) - we can justify that by saying that the moment you get `WANTED` status the ~~NSA~~ satellites are tracking you and guiding the eyebot to your current location. Still _much_ saner than before.

Addresses issue https://github.com/CleverRaven/Cataclysm-DDA/issues/215.

### CODE REVIEW

**Comment by vache** [comment]
That doesn't really address #215 at all, the eyebots still swoop down wherever you are, not where the alarm occurred.  I would just send one or two eyebots to some radius around the location where the alarm was triggered, regardless of what triggered the alarm.  Also, I'd probably cut down the time it takes to swoop in for the eyebots.  I always imagined them as part of a commercial security system thing and not a function of the police, so presumably the store or whatever alarmed building would have them in storage nearby.  That way, you have to loot fast and get out because if you're still in the building when they come, you stand a good chance of getting photographed as soon as you leave.  Also having eyebots take pictures of zombies and then a police bot swoops in and scoops up the zombie might be a fun feature, and a good way of handling Zs if you can avoid getting caught yourself.

**Comment by drbig** [comment]
@vache I agree. I also mentioned that this is the simplest fix that makes the logic much better. And I've already coded it. Feel free to overhaul the `WANTED` system so that what we both like is easy to implement.

**Comment by vache** [comment]
Looks like it might be pretty easy to do, at work or I'd do it myself, move EVENT_WANTED from per_turn to actualize, remove the one_in and make it a loop to add rng(0,2) (or however many), adjust the turn in the add_event call to something lower, like  ````int(calendar::turn) + 30```` to lower the spawn delay, then change it to where alarms can go off even if you're WANTED already.  If my understanding is correct, that should result in 0-2 eyebots swooping in around the player's location, but this time it's only 30 turns after triggering, so it's much less likely that they have escaped very far.  The part with taking picture of zombies and having the cops go after the zombies I'd have to look into more, but I think that can wait til another PR.

**Comment by drbig** [comment]
You're not implementing what we like - what we like is to have the eyebot spawn near the *trigger source* (which is non-trivial), not near the player (which is easy, and how it is now).

I can set off an alarm sitting in my vehicle and be far far away from the *trigger source* in 30 turns.

You should also note that there are more triggers that give player the `WANTED` status than just smashing windows (e.g. failing to hack a computer).

"The part with taking picture of zombies and having the cops go after the zombies" - this is definitely a separate thing.

**Comment by vache** [comment]
30 turns is just an estimate.  It could be 5 turns or 0 and not even require the event system, but that would be silly because players need SOME chance to avoid it.  The main thing is the player has a much higher chance of being in the area where the alarm was triggered.  True, the case of hitting an alarmed door with a car then speeding off would be an instance where this would be false, but they could also just bash the car through a wall and avoid setting off an alarm entirely.  Also, every single context I found WANTED being used in is an above ground building's alarmed doors or windows, whether its triggered by smashing the terrain itself or failing the computer check.

File: src/event.cpp [hunk 1]
```diff
@@ -298,6 +298,8 @@ void event::per_turn()
     g->add_zombie(eyebot);
     if (g->u.sees( place ))
      add_msg(m_warning, _("An eyebot swoops down nearby!"));
+    // One eyebot per trigger is enough, really
+    turn = int(calendar::turn);
```
Comment (KA101) [inline src/event.cpp]: OK, this looks like it messes with the timer.  Gonna look at this when I have a chance to check that system out.

**Comment by drbig** [comment]
@KA101: src/game.cpp

    1563 void game::process_events()                                                                   
    1564 {                                                                                             
    1565     for( auto it = events.begin(); it != events.end(); ) {                                    
    1566         it->per_turn();                                                                       
    1567         if (it->turn <= int(calendar::turn)) {                                                
    1568             it->actualize();                                                                  
    1569             it = events.erase(it);                                                            
    1570         } else {                                                                              
    1571             it++;                                                                             
    1572         }                                                                                     
    1573     }                                                                                         
    1574 }         

This seemed to me to be 'the best' way to tell upstream that this event is done.

**Comment by vache** [comment]
That's why I proposed moving it to actualize and just giving it a chance to spawn a random number at once.  Seemed like the simpler way of doing that.

**Comment by KA101** [comment]
Huh.  I take it that erasing the event as part of its function isn't an option?

**Comment by kevingranade** [comment]
Having an object erase itself is a very bad idea.

I'm going to pull this as an improvement.  If nothing else you don't get eyebots swooping in at you after a zombie on the other side of the map smashed a window.  But then I'm going to re-open #215 because this doesn't fix it, just makes it better in general.

**Comment by drbig** [comment]
My notes/concerns regarding implementation of a proper "spawn eyebot at trigger source" without ripping apart the `EVENT_WANTED` mechanic, as I tend to forget such details:

TL;DR: similar pains as for fixing solar panels to work independently of reality bubble; is `::spawn()` ok for off-reality bubble spawns?; seems having a general map access would help here too

 - All `g->add_event(EVENT_WANTED, ...)` calls would need to supply global absolute position of the trigger source; *or* global submap coord if we're ok with 'reduced accuracy'
 - In any case one has to be able to spawn an eyebot at the (now arbitrary) submap where the trigger happened, at a sane location; I assume we'd like a `g->m.random_outdoor_tile()` that can at least take a submap coord as an argument; I have no idea if `::spawn()` is safe for off-current-submap spawning; I assumed `g->add_zombie(eyebot)` does some sort of 'insert into reality bubble' thing (didn't read the source though)
 - The submap in question might be off reality bubble, perhaps off-loaded completely? Should we handle all cases, or not spawn if outside reality bubble/off-loaded?

At least eyebots can sanely spawn inside player's FOV as they 'swoop down from the sky'.

**Comment by BevapDin** [comment]
Documentation in the vent class:
```C++
/** Where the event happens, in global submap coordinates */
point event::map_point;
```
And `game::add_event` simply forwards the coordinates, and the event is created in map.cpp with `g->get_abs_levx(), g->get_abs_levy()`, so the coordinates are indeed global and absolute submap coordinates.

`monster::spawn` does not "work" outside the reality bubble (technically it could, but it currently creates a debugmsg when the new monster is out-of-bounds).

You can add a monster to the overmap (see `game::despawn_monster` / `overmapbuffer::despawn_monster`. It will be loaded when reality bubble shifts to encompass the monsters location. You need to give that monster invalid (because out-of-bounds) coordinates as `overmapbuffer::despawn_monster` interprets them as relative to the reality bubble. Or you make a `overmapbuffer::add_monster` function that uses absolute coordinates.

About `random_outdoor_tile`: why does the eyebot spawn outside at all? Where does it come from? It is much more plausible when it spawns adjacent to a wall or inside (under a ceiling), it can be assumed it comes out of those structures.

Alternatively: search for tinymap (in the mission_start.cpp) - it is used to load a small area (2x2 submaps) and place some mission related stuff there. The same technique can be used to place a monster there. In fact, look at `mission_start::place_jabberwock`, it properly does exactly what you want. And it allows to check for inside/outside tiles.
